diff --git a/Makefile b/Makefile
index 2791166..dab43a3 100644
--- a/Makefile
+++ b/Makefile
@@ -4,7 +4,7 @@
 
 VERSION!=git describe --long --abbrev=12 --tags --dirty 2>/dev/null || echo 0.13.0
 VPATH=doc
-PREFIX?=/usr/local
+PREFIX?=@out@/local
 BINDIR?=$(PREFIX)/bin
 SHAREDIR?=$(PREFIX)/share/aerc
 MANDIR?=$(PREFIX)/share/man
diff --git a/README.md b/README.md
index d8a84fc..41a5033 100644
--- a/README.md
+++ b/README.md
@@ -85,13 +85,13 @@ ### From Source
 
     # make install
 
-By default, aerc will install config files to directories under `/usr/local/aerc`,
+By default, aerc will install config files to directories under `@out@/local/aerc`,
 and will search for templates and stylesets in these locations in order:
 
 - `${XDG_CONFIG_HOME:-~/.config}/aerc`
 - `${XDG_DATA_HOME:-~/.local/share}/aerc`
-- `/usr/local/share/aerc`
-- `/usr/share/aerc`
+- `@out@/local/share/aerc`
+- `@out@/share/aerc`
 
 At build time it is possible to add an extra location to this list and to use
 that location as the default install location for config files by setting the
diff --git a/config/aerc.conf b/config/aerc.conf
index ebd361a..8953847 100644
--- a/config/aerc.conf
+++ b/config/aerc.conf
@@ -151,8 +151,8 @@
 #
 #   ${XDG_CONFIG_HOME:-~/.config}/aerc/stylesets
 #   ${XDG_DATA_HOME:-~/.local/share}/aerc/stylesets
-#   /usr/local/share/aerc/stylesets
-#   /usr/share/aerc/stylesets
+#   @out@/local/share/aerc/stylesets
+#   @out@/share/aerc/stylesets
 #
 #stylesets-dirs=
 
@@ -369,7 +369,7 @@
 #   ${XDG_CONFIG_HOME:-~/.config}/aerc/filters
 #   ${XDG_DATA_HOME:-~/.local/share}/aerc/filters
 #   $PREFIX/share/aerc/filters
-#   /usr/share/aerc/filters
+#   @out@/share/aerc/filters
 #
 # The following variables are defined in the filter command environment:
 #
@@ -433,8 +433,8 @@ message/rfc822=colorize
 #
 #   ${XDG_CONFIG_HOME:-~/.config}/aerc/templates
 #   ${XDG_DATA_HOME:-~/.local/share}/aerc/templates
-#   /usr/local/share/aerc/templates
-#   /usr/share/aerc/templates
+#   @out@/local/share/aerc/templates
+#   @out@/share/aerc/templates
 #
 #template-dirs=
 
diff --git a/config/config.go b/config/config.go
index c4794cc..ee19028 100644
--- a/config/config.go
+++ b/config/config.go
@@ -68,7 +68,7 @@ func buildDefaultDirs() []string {
 	}
 
 	// Add custom buildtime shareDir
-	if shareDir != "" && shareDir != "/usr/local/share/aerc" {
+	if shareDir != "" && shareDir != "@out@/local/share/aerc" {
 		shareDir, err := homedir.Expand(shareDir)
 		if err == nil {
 			defaultDirs = append(defaultDirs, shareDir)
@@ -76,8 +76,8 @@ func buildDefaultDirs() []string {
 	}
 
 	// Add fixed fallback locations
-	defaultDirs = append(defaultDirs, "/usr/local/share/aerc")
-	defaultDirs = append(defaultDirs, "/usr/share/aerc")
+	defaultDirs = append(defaultDirs, "@out@/local/share/aerc")
+	defaultDirs = append(defaultDirs, "@out@/share/aerc")
 
 	return defaultDirs
 }
diff --git a/doc/aerc-config.5.scd b/doc/aerc-config.5.scd
index 247490a..5cd2983 100644
--- a/doc/aerc-config.5.scd
+++ b/doc/aerc-config.5.scd
@@ -12,7 +12,7 @@ account credentials. We look for these files in your XDG config home plus
 _aerc_, which defaults to _~/.config/aerc_.
 
 Examples of these config files are typically included with your installation of
-aerc and are usually installed in _/usr/share/aerc_.
+aerc and are usually installed in _@out@/share/aerc_.
 
 Each file uses the ini format, and consists of sections with keys and values.
 A line beginning with _#_ is considered a comment and ignored, as are empty
@@ -277,8 +277,8 @@ These options are configured in the *[ui]* section of _aerc.conf_.
 	```
 	${XDG_CONFIG_HOME:-~/.config}/aerc/stylesets
 	${XDG_DATA_HOME:-~/.local/share}/aerc/stylesets
-	/usr/local/share/aerc/stylesets
-	/usr/share/aerc/stylesets
+	@out@/local/share/aerc/stylesets
+	@out@/share/aerc/stylesets
 	```
 
 *styleset-name*
@@ -586,7 +586,7 @@ reading from stdin by putting _-_ instead of a path to a file. You can also
 chain together multiple filters by piping with _|_.
 
 aerc ships with some default filters installed in the share directory (usually
-_/usr/share/aerc/filters_). Note that these may have additional dependencies
+_@out@/share/aerc/filters_). Note that these may have additional dependencies
 that aerc does not have alone.
 
 The filter commands are invoked with _sh -c command_. The following folders are
@@ -596,7 +596,7 @@ appended to the system *$PATH* to allow referencing filters from their name only
 ${XDG_CONFIG_HOME:-~/.config}/aerc/filters
 ${XDG_DATA_HOME:-~/.local/share}/aerc/filters
 $PREFIX/share/aerc/filters
-/usr/share/aerc/filters
+@out@/share/aerc/filters
 ```
 
 The following variables are defined in the filter command environment:
@@ -782,7 +782,7 @@ and *:forward* commands can be called with the *-T* flag with the name of the
 template name.
 
 aerc ships with some default templates installed in the share directory (usually
-_/usr/share/aerc/templates_).
+_@out@/share/aerc/templates_).
 
 These options are configured in the *[templates]* section of _aerc.conf_.
 
@@ -794,8 +794,8 @@ These options are configured in the *[templates]* section of _aerc.conf_.
 	```
 	${XDG_CONFIG_HOME:-~/.config}/aerc/templates
 	${XDG_DATA_HOME:-~/.local/share}/aerc/templates
-	/usr/local/share/aerc/templates
-	/usr/share/aerc/templates
+	@out@/local/share/aerc/templates
+	@out@/share/aerc/templates
 	```
 
 *new-message*
diff --git a/doc/aerc-sendmail.5.scd b/doc/aerc-sendmail.5.scd
index 7aa0f8a..57cf7be 100644
--- a/doc/aerc-sendmail.5.scd
+++ b/doc/aerc-sendmail.5.scd
@@ -17,7 +17,7 @@ options are available:
 	_</path/to/sendmail>_
 
 	This should be set to the path to the sendmail binary you wish to use,
-	which is generally _/usr/sbin/sendmail_. aerc will execute it with a list of
+	which is generally _@out@/sbin/sendmail_. aerc will execute it with a list of
 	recipients on the command line and pipe the message to deliver to stdin.
 
 # SEE ALSO
diff --git a/doc/aerc-templates.7.scd b/doc/aerc-templates.7.scd
index 6c9e319..6f78432 100644
--- a/doc/aerc-templates.7.scd
+++ b/doc/aerc-templates.7.scd
@@ -111,7 +111,7 @@ aerc provides the following additional functions:
 	Execute external command, provide the second argument to its stdin.
 
 	```
-	{{exec `/usr/local/share/aerc/filters/html` .OriginalText}}
+	{{exec `@out@/local/share/aerc/filters/html` .OriginalText}}
 	```
 
 *toLocal*
@@ -142,7 +142,7 @@ aerc provides the following additional functions:
 	Example: Automatic HTML parsing for text/html mime type messages
 	```
 	{{if eq .OriginalMIMEType "text/html"}}
-	{{exec `/usr/local/share/aerc/filters/html` .OriginalText | wrap 72 | quote}}
+	{{exec `@out@/local/share/aerc/filters/html` .OriginalText | wrap 72 | quote}}
 	{{else}}
 	{{wrap 72 .OriginalText | quote}}
 	{{end}}
