From 4f962144866a5e497476ff2890530db985d6094c Mon Sep 17 00:00:00 2001
From: David96 <david@hameipe.de>
Date: Thu, 14 Jul 2022 14:14:01 +0200
Subject: [PATCH] Add VideoCrop to make Zoom happy

---
 src/screencast/pipewire_screencast.c | 20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/src/screencast/pipewire_screencast.c b/src/screencast/pipewire_screencast.c
index 17f4c273..2be74509 100644
--- a/src/screencast/pipewire_screencast.c
+++ b/src/screencast/pipewire_screencast.c
@@ -294,12 +294,18 @@ static void pwr_handle_stream_param_changed(void *data, uint32_t id,
 	params[0] = build_buffer(&b, blocks, cast->screencopy_frame_info[cast->buffer_type].size,
 			cast->screencopy_frame_info[cast->buffer_type].stride, data_type);
 
-	params[1] = spa_pod_builder_add_object(&b,
+	params[1] = spa_pod_builder_add_object (
+			&b,
+			SPA_TYPE_OBJECT_ParamMeta, SPA_PARAM_Meta,
+			SPA_PARAM_META_type, SPA_POD_Id (SPA_META_VideoCrop),
+			SPA_PARAM_META_size, SPA_POD_Int (sizeof (struct spa_meta_region)));
+
+	params[2] = spa_pod_builder_add_object(&b,
 		SPA_TYPE_OBJECT_ParamMeta, SPA_PARAM_Meta,
 		SPA_PARAM_META_type, SPA_POD_Id(SPA_META_Header),
 		SPA_PARAM_META_size, SPA_POD_Int(sizeof(struct spa_meta_header)));
 
-	pw_stream_update_params(stream, params, 2);
+	pw_stream_update_params(stream, params, 3);
 }
 
 static void pwr_handle_stream_add_buffer(void *data, struct pw_buffer *buffer) {
@@ -418,6 +424,16 @@ void xdpw_pwr_enqueue_buffer(struct xdpw_screencast_instance *cast) {
 		h->dts_offset = 0;
 		logprint(TRACE, "pipewire: timestamp %"PRId64, h->pts);
 	}
+	struct spa_meta_region *video_crop;
+	video_crop = spa_buffer_find_meta_data (spa_buf, SPA_META_VideoCrop,
+				sizeof(*video_crop));
+	if (video_crop)
+	{
+		video_crop->region.position.x = 0;
+		video_crop->region.position.y = 0;
+		video_crop->region.size.width = cast->current_frame.xdpw_buffer->width;
+		video_crop->region.size.height = cast->current_frame.xdpw_buffer->height;
+	}
 
 	if (buffer_corrupt) {
 		for (uint32_t plane = 0; plane < spa_buf->n_datas; plane++) {
